src/server/trace/recorder/tracer.ts

```diff
@@ -108,23 +103,45 @@ class ContextTracer {
     await this._snapshotter.start(false);
   }
 
-  async _captureSnapshot(name: 'before' | 'after' | 'action', sdkObject: SdkObject, metadata: CallMetadata, element?: ElementHandle): Promise<void> {
+  async _captureSnapshot(name: 'before' | 'after' | 'action' | 'event', sdkObject: SdkObject, metadata: CallMetadata, element?: ElementHandle): Promise<void> {
     if (!sdkObject.attribution.page)
       return;
     const snapshotName = `${name}@${metadata.id}`;
-    snapshotsForMetadata(metadata).push({ title: name, snapshotName });
+    metadata.snapshots.push({ title: name, snapshotName });
     this._snapshotter.captureSnapshot(sdkObject.attribution.page, snapshotName, element);
   }
 
-  async onAfterCall(sdkObject: SdkObject, metadata: CallMetadata): Promise<void> {
+  onBeforeCall(sdkObject: SdkObject, metadata: CallMetadata) {
+    this._captureSnapshot('before', sdkObject, metadata);
+    this._pendingCalls.set(metadata.id, { sdkObject, metadata });
+  }
+
+  onBeforeInputAction(sdkObject: SdkObject, metadata: CallMetadata, element: ElementHandle) {
+    this._captureSnapshot('action', sdkObject, metadata, element);
+  }
+
+  onAfterCall(sdkObject: SdkObject, metadata: CallMetadata) {
+    this._captureSnapshot('after', sdkObject, metadata);
     if (!sdkObject.attribution.page)
       return;
     const event: trace.ActionTraceEvent = {
-      timestamp: monotonicTime(),
+      timestamp: metadata.startTime,
       type: 'action',
       contextId: this._contextId,
       metadata,
-      snapshots: snapshotsForMetadata(metadata),
+    };
+    this._appendTraceEvent(event);
+    this._pendingCalls.delete(metadata.id);
+  }
+
+  onEvent(sdkObject: SdkObject, metadata: CallMetadata) {
```

Comment:
We can now drop things like `trace.LoadEvent` and such.

Reply:
Not yet, I don't yet render the events on timeline...